# Week 3 - Exploratory Data Analysis

## Dataset Loading

In this section, the cleaned retail dataset is loaded for exploratory data analysis and visualization.

In [3]:
import pandas as pd
import matplotlib.pyplot as plt

df = pd.read_csv("../../data/final_cleaned_retail_dataset.csv")

df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalPrice,Year,Month,Day,DayName,Hour
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850,UNITED KINGDOM,15.30,2010,12,1,Wednesday,8
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850,UNITED KINGDOM,20.34,2010,12,1,Wednesday,8
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850,UNITED KINGDOM,22.00,2010,12,1,Wednesday,8
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850,UNITED KINGDOM,20.34,2010,12,1,Wednesday,8
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850,UNITED KINGDOM,20.34,2010,12,1,Wednesday,8


## Dataset Overview

This section explores the structure of the cleaned retail dataset, including dimensions, column names, and datatypes.

In [2]:
print(df.shape)

print(df.columns)

df.info()

(338151, 14)
Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country', 'TotalPrice', 'Year', 'Month',
       'Day', 'DayName', 'Hour'],
      dtype='object')
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 338151 entries, 0 to 338150
Data columns (total 14 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    338151 non-null  int64  
 1   StockCode    338151 non-null  object 
 2   Description  338151 non-null  object 
 3   Quantity     338151 non-null  int64  
 4   InvoiceDate  338151 non-null  object 
 5   UnitPrice    338151 non-null  float64
 6   CustomerID   338151 non-null  int64  
 7   Country      338151 non-null  object 
 8   TotalPrice   338151 non-null  float64
 9   Year         338151 non-null  int64  
 10  Month        338151 non-null  int64  
 11  Day          338151 non-null  int64  
 12  DayName      338151 non-null  object 
 13  Hour         338151

## Statistical Summary and Missing Values Analysis

Generating statistical summaries and verifying missing values after preprocessing and cleaning.

In [3]:
df.describe()

df.isnull().sum()

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
TotalPrice     0
Year           0
Month          0
Day            0
DayName        0
Hour           0
dtype: int64

## Country-wise Transaction Analysis

Analyzing transaction distribution and revenue contribution across different countries.

In [3]:
country_transactions = df['Country'].value_counts().head(10)

print(country_transactions)

Country
UNITED KINGDOM    305129
GERMANY             7463
FRANCE              6904
EIRE                5450
SPAIN               2045
BELGIUM             1660
SWITZERLAND         1433
PORTUGAL            1260
NORWAY               772
NETHERLANDS          617
Name: count, dtype: int64


In [4]:
country_revenue = df.groupby('Country')['TotalPrice'].sum().sort_values(ascending=False).head(10)

print(country_revenue)

Country
UNITED KINGDOM    3699805.854
GERMANY            138135.070
FRANCE             122037.480
EIRE               105962.060
SPAIN               29845.510
SWITZERLAND         29790.460
BELGIUM             27839.110
PORTUGAL            20462.870
NORWAY              18123.260
NETHERLANDS         15578.900
Name: TotalPrice, dtype: float64


### Observation

The United Kingdom dominates both transaction volume and total revenue within the retail dataset, contributing significantly higher sales activity compared to all other countries. Germany, France, and EIRE also show notable customer engagement and revenue generation, but at a much smaller scale compared to the United Kingdom.

## Product Performance Analysis

Identifying products with the highest sales quantities and revenue generation.

In [6]:
top_products_quantity = df.groupby('Description')['Quantity'].sum().sort_values(ascending=False).head(10)

print(top_products_quantity)

Description
PACK OF 72 RETROSPOT CAKE CASES       15009
ASSORTED COLOUR BIRD ORNAMENT         13673
JUMBO BAG RED RETROSPOT               12170
WHITE HANGING HEART T-LIGHT HOLDER    10803
LUNCH BAG RED RETROSPOT                9797
JAM MAKING SET PRINTED                 8687
PACK OF 60 PINK PAISLEY CAKE CASES     8532
60 TEATIME FAIRY CAKE CASES            8401
VICTORIAN GLASS HANGING T-LIGHT        7630
RED RETROSPOT WRAP                     7472
Name: Quantity, dtype: int64


In [7]:
top_products_revenue = df.groupby('Description')['TotalPrice'].sum().sort_values(ascending=False).head(10)

print(top_products_revenue)

Description
PARTY BUNTING                         35011.95
WHITE HANGING HEART T-LIGHT HOLDER    31819.81
JUMBO BAG RED RETROSPOT               24929.74
SPOTTY BUNTING                        23542.75
ASSORTED COLOUR BIRD ORNAMENT         23107.37
JAM MAKING SET WITH JARS              21184.00
HOT WATER BOTTLE KEEP CALM            20741.10
SET OF 3 CAKE TINS PANTRY DESIGN      20633.47
CHOCOLATE HOT WATER BOTTLE            18462.45
CHILLI LIGHTS                         18401.20
Name: TotalPrice, dtype: float64


### Observation

Products such as “PACK OF 72 RETROSPOT CAKE CASES” and “ASSORTED COLOUR BIRD ORNAMENT” recorded the highest sales quantities, indicating strong customer demand and frequent purchases. In terms of revenue generation, products like “PARTY BUNTING” and “WHITE HANGING HEART T-LIGHT HOLDER” contributed the highest overall sales value, highlighting their strong commercial performance within the retail dataset.

In [2]:
print(df.groupby('Country')['InvoiceNo'].nunique().sort_values(ascending=False).head(10))

Country
UNITED KINGDOM    15144
GERMANY             425
FRANCE              362
EIRE                238
BELGIUM              93
SPAIN                82
NETHERLANDS          65
PORTUGAL             46
AUSTRALIA            44
SWITZERLAND          43
Name: InvoiceNo, dtype: int64


### Observation

The United Kingdom recorded the highest number of unique transactions, indicating significantly stronger customer purchasing activity compared to other countries. Germany, France, and EIRE also contributed notable transaction volumes, while the remaining countries showed comparatively lower retail activity.

## Monthly Revenue Trend Analysis

Analyzing monthly sales revenue and transaction trends to identify periods of higher customer purchasing activity.

In [3]:
monthly_revenue = df.groupby('Month')['TotalPrice'].sum().sort_values(ascending=False)

print(monthly_revenue)

monthly_transactions = df.groupby('Month')['InvoiceNo'].nunique()

print(monthly_transactions)

Month
11    647874.300
10    527605.080
9     482556.772
12    464311.210
5     325996.840
8     306043.760
6     291135.140
7     289833.551
3     286501.840
4     242996.611
1     233908.010
2     220099.410
Name: TotalPrice, dtype: float64
Month
1      904
2      895
3     1198
4     1034
5     1389
6     1268
7     1213
8     1150
9     1597
10    1741
11    2448
12    1989
Name: InvoiceNo, dtype: int64


## Weekday-wise Customer Activity Analysis

Analyzing customer purchasing behavior across different weekdays.

In [4]:
day_transactions = df['DayName'].value_counts()

print(day_transactions)

day_revenue = df.groupby('DayName')['TotalPrice'].sum().sort_values(ascending=False)

print(day_revenue)

DayName
Thursday     66856
Wednesday    57643
Tuesday      56070
Sunday       55925
Monday       55221
Friday       46436
Name: count, dtype: int64
DayName
Thursday     930363.220
Wednesday    775215.240
Tuesday      746821.181
Monday       704617.991
Friday       649834.091
Sunday       512010.801
Name: TotalPrice, dtype: float64


## Hour-wise Transaction Analysis

Analyzing transaction distribution across different hours to identify peak shopping periods.

In [2]:
hourly_transactions = df['Hour'].value_counts().sort_index()

print(hourly_transactions)

hourly_revenue = df.groupby('Hour')['TotalPrice'].sum().sort_values(ascending=False)

print(hourly_revenue)

Hour
6         1
7       178
8      6965
9     17566
10    30440
11    41621
12    62121
13    55446
14    46876
15    38849
16    20856
17    11219
18     2415
19     2879
20      719
Name: count, dtype: int64
Hour
12    721416.950
13    645327.290
14    562130.941
10    513356.131
11    503657.690
15    476431.050
9     339472.261
16    227083.190
8     139670.660
17    124019.851
18     32076.270
19     25140.010
20      4545.670
7       4530.310
6          4.250
Name: TotalPrice, dtype: float64


### Observation

The analysis shows that November generated the highest overall revenue and transaction volume, indicating a peak sales period within the retail dataset. Thursday recorded the highest customer activity and revenue contribution among all weekdays, suggesting increased purchasing behavior during that period.

Hourly transaction analysis reveals that customer activity and revenue generation peak during midday hours, particularly between 12 PM and 2 PM, highlighting the busiest shopping window within the dataset.

## Customer Purchase Analysis

Analyzing customer purchasing frequency and identifying top customers based on transaction activity and revenue contribution.

In [4]:
top_customers_transactions = df.groupby('CustomerID')['InvoiceNo'].nunique().sort_values(ascending=False).head(10)

print(top_customers_transactions)

CustomerID
12748    196
14911    193
17841    123
15311     91
14606     90
13089     80
12971     78
14527     54
13408     52
14646     46
Name: InvoiceNo, dtype: int64


In [5]:
top_customers_revenue = df.groupby('CustomerID')['TotalPrice'].sum().sort_values(ascending=False).head(10)

print(top_customers_revenue)

CustomerID
14911    85018.78
13089    36411.63
14096    36019.84
17841    32644.72
14298    24387.53
12748    21046.63
14156    18300.66
13081    17925.53
15311    17756.31
14088    15700.62
Name: TotalPrice, dtype: float64


### Observation

The customer analysis reveals that a small group of customers contributes a significantly higher number of transactions and overall revenue compared to others. Customer ID 14911 generated the highest total revenue, while Customer ID 12748 recorded the highest transaction frequency, indicating strong repeat purchasing behavior among selected customers within the retail dataset.

## Quantity and Pricing Analysis

Analyzing product quantity distribution and unit price behavior across retail transactions.

In [6]:
print(df['Quantity'].describe())

count    338151.000000
mean          7.476917
std           6.770795
min           1.000000
25%           2.000000
50%           6.000000
75%          12.000000
max          27.000000
Name: Quantity, dtype: float64


In [7]:
print(df['UnitPrice'].describe())

count    338151.000000
mean          2.192017
std           1.544770
min           0.001000
25%           1.250000
50%           1.650000
75%           2.950000
max           7.500000
Name: UnitPrice, dtype: float64


In [8]:
highest_priced_products = df[['Description', 'UnitPrice']].sort_values(by='UnitPrice', ascending=False).head(10)

print(highest_priced_products)

                        Description  UnitPrice
142531     PARISIENNE CURIO CABINET        7.5
118118     PARISIENNE CURIO CABINET        7.5
259990     PARISIENNE CURIO CABINET        7.5
142532  PARISIENNE JEWELLERY DRAWER        7.5
284343        WHITE WIRE EGG HOLDER        7.5
184860        WHITE WIRE EGG HOLDER        7.5
319112        WHITE WIRE EGG HOLDER        7.5
292599     PARISIENNE CURIO CABINET        7.5
226623     PARISIENNE CURIO CABINET        7.5
138960     PARISIENNE CURIO CABINET        7.5


### Observation

The quantity analysis shows that most retail transactions involve relatively small purchase quantities, with an average quantity of approximately 7 units per transaction. The unit price analysis indicates that the majority of products are moderately priced, while a few premium products such as “PARISIENNE CURIO CABINET” and “WHITE WIRE EGG HOLDER” recorded the highest unit prices within the cleaned dataset.